Python Enum support for click.Choice

I can work  around it as [described here](https://pythonpath.wordpress.com/2012/01/26/python-enum-with-namedtuple/) it would be great if `Choice` supported Python Enum.

Unless of course I have missed something completely fundamental :-)

Well, 
```
$ cat click_test.py
import click
import enum


class MyEnum(str, enum.Enum):
    A = "a"


@click.command("my-cmd")
@click.option("which_option", type=click.Choice(MyEnum))
def my_cmd(which_option: MyEnum):
    print(which_option)
    print(type(which_option))

    reveal_type(MyEnum)


if __name__ == "__main__":
    my_cmd() 
```

Still Gives:
```
click_test.py:10: error: Argument 1 to "Choice" has incompatible type "Type[MyEnum]"; expected "Sequence[str]"
click_test.py:15: note: Revealed type is "Overload(def (object: builtins.object =) -> click_test.MyEnum*, def (o: builtins.bytes, encoding: builtins.str =, errors: builtins.str =) -> click_test.MyEnum*)"
```

--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -96,7 +96,8 @@ Unreleased
     ``Option.default`` not needing to implement `__bool__`. :pr:`2829`
 -   Incorrect ``click.edit`` typing has been corrected. :pr:`2804`
 -   :class:``Choice`` is now generic and supports any iterable value.
-    This allows you to use enums and other non-``str`` values.
+    This allows you to use enums and other non-``str`` values. :pr:`2796`
+    :issue:`605`
 
 Version 8.1.8
 -------------

--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -88,13 +88,15 @@ Unreleased
     -   Parameters cannot be required nor prompted or an error is raised.
     -   A warning will be printed when something deprecated is used.
 
--   Add a ``catch_exceptions`` parameter to :class:`CliRunner`. If
-    ``catch_exceptions`` is not passed to :meth:`CliRunner.invoke`,
+-   Add a ``catch_exceptions`` parameter to :class:``CliRunner``. If
+    ``catch_exceptions`` is not passed to :meth:``CliRunner.invoke``,
     the value from :class:`CliRunner`. :issue:`2817` :pr:`2818`
 -   ``Option.flag_value`` will no longer have a default value set based on
     ``Option.default`` if ``Option.is_flag`` is ``False``. This results in
     ``Option.default`` not needing to implement `__bool__`. :pr:`2829`
 -   Incorrect ``click.edit`` typing has been corrected. :pr:`2804`
+-   :class:``Choice`` is now generic and supports any iterable value.
+    This allows you to use enums and other non-``str`` values.
 
 Version 8.1.8
 -------------

--- a/docs/options.rst
+++ b/docs/options.rst
@@ -381,7 +381,9 @@ Example:
 
 .. click:example::
 
-    class HashType:
+    import enum
+
+    class HashType(enum.Enum):
         MD5 = 'MD5'
         SHA1 = 'SHA1'
 
@@ -411,8 +413,8 @@ Choices work with options that have ``multiple=True``. If a ``default``
 value is given with ``multiple=True``, it should be a list or tuple of
 valid choices.
 
-Choices should be unique after considering the effects of
-``case_sensitive`` and any specified token normalization function.
+Choices should be unique after normalization, see
+:meth:`Choice.normalize_choice` for more info.
 
 .. versionchanged:: 7.1
     The resulting value from an option will always be one of the

--- a/docs/options.rst
+++ b/docs/options.rst
@@ -375,15 +375,20 @@ In that case you can use :class:`Choice` type.  It can be instantiated
 with a list of valid values.  The originally passed choice will be returned,
 not the str passed on the command line.  Token normalization functions and
 ``case_sensitive=False`` can cause the two to be different but still match.
+:meth:`Choice.normalize_choice` for more info.
 
 Example:
 
 .. click:example::
 
+    class HashType:
+        MD5 = 'MD5'
+        SHA1 = 'SHA1'
+
     @click.command()
     @click.option('--hash-type',
-                  type=click.Choice(['MD5', 'SHA1'], case_sensitive=False))
-    def digest(hash_type):
+                  type=click.Choice(HashType, case_sensitive=False))
+    def digest(hash_type: HashType):
         click.echo(hash_type)
 
 What it looks like:
@@ -398,8 +403,9 @@ What it looks like:
     println()
     invoke(digest, args=['--help'])
 
-Only pass the choices as list or tuple. Other iterables (like
-generators) may lead to unexpected results.
+Since version 8.2.0 any iterable may be passed to :class:`Choice`, here
+an ``Enum`` is used which will result in all enum values to be valid
+choices.
 
 Choices work with options that have ``multiple=True``. If a ``default``
 value is given with ``multiple=True``, it should be a list or tuple of

--- a/src/click/core.py
+++ b/src/click/core.py
@@ -2192,11 +2192,11 @@ class Parameter:
         """
         return self.name  # type: ignore
 
-    def make_metavar(self) -> str:
+    def make_metavar(self, ctx: Context) -> str:
         if self.metavar is not None:
             return self.metavar
 
-        metavar = self.type.get_metavar(self)
+        metavar = self.type.get_metavar(param=self, ctx=ctx)
 
         if metavar is None:
             metavar = self.type.name.upper()
@@ -2775,7 +2775,7 @@ class Option(Parameter):
                 any_prefix_is_slash = True
 
             if not self.is_flag and not self.count:
-                rv += f" {self.make_metavar()}"
+                rv += f" {self.make_metavar(ctx=ctx)}"
 
             return rv
 
@@ -3056,10 +3056,10 @@ class Argument(Parameter):
             return self.metavar
         return self.name.upper()  # type: ignore
 
-    def make_metavar(self) -> str:
+    def make_metavar(self, ctx: Context) -> str:
         if self.metavar is not None:
             return self.metavar
-        var = self.type.get_metavar(self)
+        var = self.type.get_metavar(param=self, ctx=ctx)
         if not var:
             var = self.name.upper()  # type: ignore
         if self.deprecated:
@@ -3088,10 +3088,10 @@ class Argument(Parameter):
         return name, [arg], []
 
     def get_usage_pieces(self, ctx: Context) -> list[str]:
-        return [self.make_metavar()]
+        return [self.make_metavar(ctx)]
 
     def get_error_hint(self, ctx: Context) -> str:
-        return f"'{self.make_metavar()}'"
+        return f"'{self.make_metavar(ctx)}'"
 
     def add_to_parser(self, parser: _OptionParser, ctx: Context) -> None:
         parser.add_argument(dest=self.name, nargs=self.nargs, obj=self)

--- a/src/click/exceptions.py
+++ b/src/click/exceptions.py
@@ -174,7 +174,9 @@ class MissingParameter(BadParameter):
 
         msg = self.message
         if self.param is not None:
-            msg_extra = self.param.type.get_missing_message(self.param)
+            msg_extra = self.param.type.get_missing_message(
+                param=self.param, ctx=self.ctx
+            )
             if msg_extra:
                 if msg:
                     msg += f". {msg_extra}"

--- a/src/click/types.py
+++ b/src/click/types.py
@@ -23,6 +23,8 @@ if t.TYPE_CHECKING:
     from .core import Parameter
     from .shell_completion import CompletionItem
 
+ParamTypeValue = t.TypeVar("ParamTypeValue")
+
 
 class ParamType:
     """Represents the type of a parameter. Validates and converts values
@@ -86,10 +88,10 @@ class ParamType:
         if value is not None:
             return self.convert(value, param, ctx)
 
-    def get_metavar(self, param: Parameter) -> str | None:
+    def get_metavar(self, param: Parameter, ctx: Context) -> str | None:
         """Returns the metavar default for this param if it provides one."""
 
-    def get_missing_message(self, param: Parameter) -> str | None:
+    def get_missing_message(self, param: Parameter, ctx: Context | None) -> str | None:
         """Optionally might return extra information about a missing
         parameter.
 
@@ -227,7 +229,7 @@ class StringParamType(ParamType):
         return "STRING"
 
 
-class Choice(ParamType):
+class Choice(ParamType, t.Generic[ParamTypeValue]):
     """The choice type allows a value to be checked against a fixed set
     of supported values. All of these values have to be strings.
 
@@ -247,7 +249,7 @@ class Choice(ParamType):
     name = "choice"
 
     def __init__(
-        self, choices: cabc.Sequence[str], case_sensitive: bool = True
+        self, choices: cabc.Sequence[ParamTypeValue], case_sensitive: bool = True
     ) -> None:
         self.choices = choices
         self.case_sensitive = case_sensitive
@@ -258,14 +260,52 @@ class Choice(ParamType):
         info_dict["case_sensitive"] = self.case_sensitive
         return info_dict
 
-    def get_metavar(self, param: Parameter) -> str:
+    def normalized_mapping(
+        self, ctx: Context | None = None
+    ) -> cabc.Mapping[ParamTypeValue, str]:
+        """
+        Returns mapping where keys are the original choices and the values are
+        the normalized values that are accepted via the command line.
+
+        .. versionadded:: 8.2.0
+        """
+        return {
+            choice: self.normalize_choice(
+                choice=choice,
+                ctx=ctx,
+            )
+            for choice in self.choices
+        }
+
+    def normalize_choice(self, choice: ParamTypeValue, ctx: Context | None) -> str:
+        """
+        Normalize a choice value.
+
+        By default use ``ctx.token_normalize_func`` and if not case sensitive,
+        convert to a lowecase/casefolded value.
+
+        .. versionadded:: 8.2.0
+        """
+        normed_value = str(choice)
+
+        if ctx is not None and ctx.token_normalize_func is not None:
+            normed_value = ctx.token_normalize_func(normed_value)
+
+        if not self.case_sensitive:
+            normed_value = normed_value.casefold()
+
+        return normed_value
+
+    def get_metavar(self, param: Parameter, ctx: Context) -> str | None:
         if param.param_type_name == "option" and not param.show_choices:  # type: ignore
             choice_metavars = [
                 convert_type(type(choice)).name.upper() for choice in self.choices
             ]
             choices_str = "|".join([*dict.fromkeys(choice_metavars)])
         else:
-            choices_str = "|".join([str(i) for i in self.choices])
+            choices_str = "|".join(
+                [str(i) for i in self.normalized_mapping(ctx=ctx).values()]
+            )
 
         # Use curly braces to indicate a required argument.
         if param.required and param.param_type_name == "argument":
@@ -274,51 +314,56 @@ class Choice(ParamType):
         # Use square braces to indicate an option or optional argument.
         return f"[{choices_str}]"
 
-    def get_missing_message(self, param: Parameter) -> str:
-        return _("Choose from:\n\t{choices}").format(choices=",\n\t".join(self.choices))
+    def get_missing_message(self, param: Parameter, ctx: Context | None) -> str:
+        """
+        Message shown when no choice is passed.
+
+        .. versionchanged:: 8.2.0 Added ``ctx`` argument.
+        """
+        return _("Choose from:\n\t{choices}").format(
+            choices=",\n\t".join(self.normalized_mapping(ctx=ctx).values())
+        )
 
     def convert(
         self, value: t.Any, param: Parameter | None, ctx: Context | None
     ) -> t.Any:
-        # Match through normalization and case sensitivity
-        # first do token_normalize_func, then lowercase
-        # preserve original `value` to produce an accurate message in
-        # `self.fail`
-        normed_value = value
-        normed_choices = {choice: choice for choice in self.choices}
+        normed_value = self.normalize_choice(choice=value, ctx=ctx)
+        normalized_mapping = self.normalized_mapping(ctx=ctx)
+        original_choice = next(
+            (
+                original
+                for original, normalized in normalized_mapping.items()
+                if normalized == normed_value
+            ),
+            None,
+        )
 
-        if ctx is not None and ctx.token_normalize_func is not None:
-            normed_value = ctx.token_normalize_func(value)
-            normed_choices = {
-                ctx.token_normalize_func(normed_choice): original
-                for normed_choice, original in normed_choices.items()
-            }
+        if not original_choice:
+            self.fail(
+                self.get_invalid_choice_message(value=value, ctx=ctx),
+                param=param,
+                ctx=ctx,
+            )
 
-        if not self.case_sensitive:
-            normed_value = normed_value.casefold()
-            normed_choices = {
-                normed_choice.casefold(): original
-                for normed_choice, original in normed_choices.items()
-            }
+        return original_choice
 
-        if normed_value in normed_choices:
-            return normed_choices[normed_value]
-
-        self.fail(self.get_invalid_choice_message(value), param, ctx)
-
-    def get_invalid_choice_message(self, value: t.Any) -> str:
+    def get_invalid_choice_message(self, value: t.Any, ctx: Context | None) -> str:
         """Get the error message when the given choice is invalid.
 
         :param value: The invalid value.
 
         .. versionadded:: 8.2
         """
-        choices_str = ", ".join(map(repr, self.choices))
-        return ngettext(
-            "{value!r} is not {choice}.",
-            "{value!r} is not one of {choices}.",
-            len(self.choices),
-        ).format(value=value, choice=choices_str, choices=choices_str)
+        choices_str = ", ".join(self.normalized_mapping(ctx=ctx).values())
+        self.fail(
+            message=ngettext(
+                "{value!r} is not {choice}.",
+                "{value!r} is not one of {choices}.",
+                len(self.choices),
+            ).format(value=value, choice=choices_str, choices=choices_str),
+            param=value,
+            ctx=ctx,
+        )
 
     def __repr__(self) -> str:
         return f"Choice({list(self.choices)})"
@@ -382,7 +427,7 @@ class DateTime(ParamType):
         info_dict["formats"] = self.formats
         return info_dict
 
-    def get_metavar(self, param: Parameter) -> str:
+    def get_metavar(self, param: Parameter, ctx: Context) -> str | None:
         return f"[{'|'.join(self.formats)}]"
 
     def _try_to_convert_date(self, value: t.Any, format: str) -> datetime | None:

--- a/src/click/types.py
+++ b/src/click/types.py
@@ -250,7 +250,7 @@ class Choice(ParamType, t.Generic[ParamTypeValue]):
         tuple.
 
     .. versionadded:: 8.2.0
-        Choices are normalized via :meth:`normalize_choice`
+        Choice normalization can be overridden via :meth:`normalize_choice`.
     """
 
     name = "choice"

--- a/src/click/types.py
+++ b/src/click/types.py
@@ -1,6 +1,7 @@
 from __future__ import annotations
 
 import collections.abc as cabc
+import enum
 import os
 import stat
 import sys
@@ -249,9 +250,9 @@ class Choice(ParamType, t.Generic[ParamTypeValue]):
     name = "choice"
 
     def __init__(
-        self, choices: cabc.Sequence[ParamTypeValue], case_sensitive: bool = True
+        self, choices: cabc.Iterable[ParamTypeValue], case_sensitive: bool = True
     ) -> None:
-        self.choices = choices
+        self.choices: cabc.Sequence[ParamTypeValue] = list(choices)
         self.case_sensitive = case_sensitive
 
     def to_info_dict(self) -> dict[str, t.Any]:
@@ -277,7 +278,7 @@ class Choice(ParamType, t.Generic[ParamTypeValue]):
             for choice in self.choices
         }
 
-    def normalize_choice(self, choice: ParamTypeValue, ctx: Context | None) -> str:
+    def normalize_choice(self, choice: t.Any, ctx: Context | None) -> str:
         """
         Normalize a choice value.
 
@@ -286,7 +287,7 @@ class Choice(ParamType, t.Generic[ParamTypeValue]):
 
         .. versionadded:: 8.2.0
         """
-        normed_value = str(choice)
+        normed_value = choice.name if isinstance(choice, enum.Enum) else str(choice)
 
         if ctx is not None and ctx.token_normalize_func is not None:
             normed_value = ctx.token_normalize_func(normed_value)
@@ -326,26 +327,27 @@ class Choice(ParamType, t.Generic[ParamTypeValue]):
 
     def convert(
         self, value: t.Any, param: Parameter | None, ctx: Context | None
-    ) -> t.Any:
+    ) -> ParamTypeValue:
+        """
+        For a given value from the parser, normalize it and find its
+        matching normalized value in the list of choices. Then return the
+        matched "original" choice.
+        """
         normed_value = self.normalize_choice(choice=value, ctx=ctx)
         normalized_mapping = self.normalized_mapping(ctx=ctx)
-        original_choice = next(
-            (
+
+        try:
+            return next(
                 original
                 for original, normalized in normalized_mapping.items()
                 if normalized == normed_value
-            ),
-            None,
-        )
-
-        if not original_choice:
+            )
+        except StopIteration:
             self.fail(
                 self.get_invalid_choice_message(value=value, ctx=ctx),
                 param=param,
                 ctx=ctx,
             )
-
-        return original_choice
 
     def get_invalid_choice_message(self, value: t.Any, ctx: Context | None) -> str:
         """Get the error message when the given choice is invalid.

--- a/src/click/types.py
+++ b/src/click/types.py
@@ -232,19 +232,25 @@ class StringParamType(ParamType):
 
 class Choice(ParamType, t.Generic[ParamTypeValue]):
     """The choice type allows a value to be checked against a fixed set
-    of supported values. All of these values have to be strings.
+    of supported values.
 
-    You should only pass a list or tuple of choices. Other iterables
-    (like generators) may lead to surprising results.
+    You may pass any iterable value which will be converted to a tuple
+    and thus will only be iterated once.
 
-    The resulting value will always be one of the originally passed choices
-    regardless of ``case_sensitive`` or any ``ctx.token_normalize_func``
-    being specified.
-
-    See :ref:`choice-opts` for an example.
+    The resulting value will always be one of the originally passed choices.
+    See :meth:`normalize_choice` for more info on the mapping of strings
+    to choices. See :ref:`choice-opts` for an example.
 
     :param case_sensitive: Set to false to make choices case
         insensitive. Defaults to true.
+
+    .. versionchanged:: 8.2.0
+        Non-``str`` ``choices`` are now supported. It can additionally be any
+        iterable. Before you were not recommended to pass anything but a list or
+        tuple.
+
+    .. versionadded:: 8.2.0
+        Choices are normalized via :meth:`normalize_choice`
     """
 
     name = "choice"
@@ -252,7 +258,7 @@ class Choice(ParamType, t.Generic[ParamTypeValue]):
     def __init__(
         self, choices: cabc.Iterable[ParamTypeValue], case_sensitive: bool = True
     ) -> None:
-        self.choices: cabc.Sequence[ParamTypeValue] = list(choices)
+        self.choices: cabc.Sequence[ParamTypeValue] = tuple(choices)
         self.case_sensitive = case_sensitive
 
     def to_info_dict(self) -> dict[str, t.Any]:
@@ -261,14 +267,15 @@ class Choice(ParamType, t.Generic[ParamTypeValue]):
         info_dict["case_sensitive"] = self.case_sensitive
         return info_dict
 
-    def normalized_mapping(
+    def _normalized_mapping(
         self, ctx: Context | None = None
     ) -> cabc.Mapping[ParamTypeValue, str]:
         """
         Returns mapping where keys are the original choices and the values are
         the normalized values that are accepted via the command line.
 
-        .. versionadded:: 8.2.0
+        This is a simple wrapper around :meth:`normalize_choice`, use that
+        instead which is supported.
         """
         return {
             choice: self.normalize_choice(
@@ -278,12 +285,13 @@ class Choice(ParamType, t.Generic[ParamTypeValue]):
             for choice in self.choices
         }
 
-    def normalize_choice(self, choice: t.Any, ctx: Context | None) -> str:
+    def normalize_choice(self, choice: ParamTypeValue, ctx: Context | None) -> str:
         """
-        Normalize a choice value.
+        Normalize a choice value, used to map a passed string to a choice.
+        Each choice must have a unique normalized value.
 
-        By default use ``ctx.token_normalize_func`` and if not case sensitive,
-        convert to a lowecase/casefolded value.
+        By default uses :meth:`Context.token_normalize_func` and if not case
+        sensitive, convert it to a casefolded value.
 
         .. versionadded:: 8.2.0
         """
@@ -305,7 +313,7 @@ class Choice(ParamType, t.Generic[ParamTypeValue]):
             choices_str = "|".join([*dict.fromkeys(choice_metavars)])
         else:
             choices_str = "|".join(
-                [str(i) for i in self.normalized_mapping(ctx=ctx).values()]
+                [str(i) for i in self._normalized_mapping(ctx=ctx).values()]
             )
 
         # Use curly braces to indicate a required argument.
@@ -322,7 +330,7 @@ class Choice(ParamType, t.Generic[ParamTypeValue]):
         .. versionchanged:: 8.2.0 Added ``ctx`` argument.
         """
         return _("Choose from:\n\t{choices}").format(
-            choices=",\n\t".join(self.normalized_mapping(ctx=ctx).values())
+            choices=",\n\t".join(self._normalized_mapping(ctx=ctx).values())
         )
 
     def convert(
@@ -334,7 +342,7 @@ class Choice(ParamType, t.Generic[ParamTypeValue]):
         matched "original" choice.
         """
         normed_value = self.normalize_choice(choice=value, ctx=ctx)
-        normalized_mapping = self.normalized_mapping(ctx=ctx)
+        normalized_mapping = self._normalized_mapping(ctx=ctx)
 
         try:
             return next(
@@ -356,7 +364,7 @@ class Choice(ParamType, t.Generic[ParamTypeValue]):
 
         .. versionadded:: 8.2
         """
-        choices_str = ", ".join(map(repr, self.normalized_mapping(ctx=ctx).values()))
+        choices_str = ", ".join(map(repr, self._normalized_mapping(ctx=ctx).values()))
         return ngettext(
             "{value!r} is not {choice}.",
             "{value!r} is not one of {choices}.",

--- a/src/click/types.py
+++ b/src/click/types.py
@@ -354,16 +354,12 @@ class Choice(ParamType, t.Generic[ParamTypeValue]):
 
         .. versionadded:: 8.2
         """
-        choices_str = ", ".join(self.normalized_mapping(ctx=ctx).values())
-        self.fail(
-            message=ngettext(
-                "{value!r} is not {choice}.",
-                "{value!r} is not one of {choices}.",
-                len(self.choices),
-            ).format(value=value, choice=choices_str, choices=choices_str),
-            param=value,
-            ctx=ctx,
-        )
+        choices_str = ", ".join(map(repr, self.normalized_mapping(ctx=ctx).values()))
+        return ngettext(
+            "{value!r} is not {choice}.",
+            "{value!r} is not one of {choices}.",
+            len(self.choices),
+        ).format(value=value, choice=choices_str, choices=choices_str)
 
     def __repr__(self) -> str:
         return f"Choice({list(self.choices)})"

--- a/tests/test_basic.py
+++ b/tests/test_basic.py
@@ -382,6 +382,32 @@ def test_choice_option(runner):
     assert "--method [foo|bar|baz]" in result.output
 
 
+def test_choice_option_normalization(runner):
+    @click.command()
+    @click.option(
+        "--method",
+        type=click.Choice(
+            ["SCREAMING_SNAKE_CASE", "snake_case", "PascalCase", "kebab-case"]
+        ),
+    )
+    def cli(method):
+        click.echo(method)
+
+    result = runner.invoke(cli, ["--method=foo"])
+    assert not result.exception
+    assert result.output == "foo\n"
+
+    result = runner.invoke(cli, ["--method=meh"])
+    assert result.exit_code == 2
+    assert (
+        "Invalid value for '--method': 'meh' is not one of "
+        "'screaming-snake-case', 'snake-case', 'pascalcase', 'kebab-case'."
+    ) in result.output
+
+    result = runner.invoke(cli, ["--help"])
+    assert "--method [foo|bar|baz]" in result.output
+
+
 def test_choice_argument(runner):
     @click.command()
     @click.argument("method", type=click.Choice(["foo", "bar", "baz"]))

--- a/tests/test_basic.py
+++ b/tests/test_basic.py
@@ -387,25 +387,34 @@ def test_choice_option_normalization(runner):
     @click.option(
         "--method",
         type=click.Choice(
-            ["SCREAMING_SNAKE_CASE", "snake_case", "PascalCase", "kebab-case"]
+            ["SCREAMING_SNAKE_CASE", "snake_case", "PascalCase", "kebab-case"],
+            case_sensitive=False,
         ),
     )
     def cli(method):
         click.echo(method)
 
-    result = runner.invoke(cli, ["--method=foo"])
-    assert not result.exception
-    assert result.output == "foo\n"
+    result = runner.invoke(cli, ["--method=snake_case"])
+    assert not result.exception, result.output
+    assert result.output == "snake_case\n"
+
+    # Even though it's case sensitive, the choice's original value is preserved
+    result = runner.invoke(cli, ["--method=pascalcase"])
+    assert not result.exception, result.output
+    assert result.output == "PascalCase\n"
 
     result = runner.invoke(cli, ["--method=meh"])
     assert result.exit_code == 2
     assert (
         "Invalid value for '--method': 'meh' is not one of "
-        "'screaming-snake-case', 'snake-case', 'pascalcase', 'kebab-case'."
+        "'screaming_snake_case', 'snake_case', 'pascalcase', 'kebab-case'."
     ) in result.output
 
     result = runner.invoke(cli, ["--help"])
-    assert "--method [foo|bar|baz]" in result.output
+    assert (
+        "--method [screaming_snake_case|snake_case|pascalcase|kebab-case]"
+        in result.output
+    )
 
 
 def test_choice_argument(runner):

--- a/tests/test_basic.py
+++ b/tests/test_basic.py
@@ -1,3 +1,6 @@
+from __future__ import annotations
+
+import enum
 import os
 from itertools import chain
 
@@ -382,41 +385,6 @@ def test_choice_option(runner):
     assert "--method [foo|bar|baz]" in result.output
 
 
-def test_choice_option_normalization(runner):
-    @click.command()
-    @click.option(
-        "--method",
-        type=click.Choice(
-            ["SCREAMING_SNAKE_CASE", "snake_case", "PascalCase", "kebab-case"],
-            case_sensitive=False,
-        ),
-    )
-    def cli(method):
-        click.echo(method)
-
-    result = runner.invoke(cli, ["--method=snake_case"])
-    assert not result.exception, result.output
-    assert result.output == "snake_case\n"
-
-    # Even though it's case sensitive, the choice's original value is preserved
-    result = runner.invoke(cli, ["--method=pascalcase"])
-    assert not result.exception, result.output
-    assert result.output == "PascalCase\n"
-
-    result = runner.invoke(cli, ["--method=meh"])
-    assert result.exit_code == 2
-    assert (
-        "Invalid value for '--method': 'meh' is not one of "
-        "'screaming_snake_case', 'snake_case', 'pascalcase', 'kebab-case'."
-    ) in result.output
-
-    result = runner.invoke(cli, ["--help"])
-    assert (
-        "--method [screaming_snake_case|snake_case|pascalcase|kebab-case]"
-        in result.output
-    )
-
-
 def test_choice_argument(runner):
     @click.command()
     @click.argument("method", type=click.Choice(["foo", "bar", "baz"]))
@@ -436,6 +404,82 @@ def test_choice_argument(runner):
 
     result = runner.invoke(cli, ["--help"])
     assert "{foo|bar|baz}" in result.output
+
+
+def test_choice_argument_enum(runner):
+    class MyEnum(str, enum.Enum):
+        FOO = "foo-value"
+        BAR = "bar-value"
+        BAZ = "baz-value"
+
+    @click.command()
+    @click.argument("method", type=click.Choice(MyEnum, case_sensitive=False))
+    def cli(method: MyEnum):
+        assert isinstance(method, MyEnum)
+        click.echo(method)
+
+    result = runner.invoke(cli, ["foo"])
+    assert result.output == "foo-value\n"
+    assert not result.exception
+
+    result = runner.invoke(cli, ["meh"])
+    assert result.exit_code == 2
+    assert (
+        "Invalid value for '{foo|bar|baz}': 'meh' is not one of 'foo',"
+        " 'bar', 'baz'." in result.output
+    )
+
+    result = runner.invoke(cli, ["--help"])
+    assert "{foo|bar|baz}" in result.output
+
+
+def test_choice_argument_custom_type(runner):
+    class MyClass:
+        def __init__(self, value: str) -> None:
+            self.value = value
+
+        def __str__(self) -> str:
+            return self.value
+
+    @click.command()
+    @click.argument(
+        "method", type=click.Choice([MyClass("foo"), MyClass("bar"), MyClass("baz")])
+    )
+    def cli(method: MyClass):
+        assert isinstance(method, MyClass)
+        click.echo(method)
+
+    result = runner.invoke(cli, ["foo"])
+    assert not result.exception
+    assert result.output == "foo\n"
+
+    result = runner.invoke(cli, ["meh"])
+    assert result.exit_code == 2
+    assert (
+        "Invalid value for '{foo|bar|baz}': 'meh' is not one of 'foo',"
+        " 'bar', 'baz'." in result.output
+    )
+
+    result = runner.invoke(cli, ["--help"])
+    assert "{foo|bar|baz}" in result.output
+
+
+def test_choice_argument_none(runner):
+    @click.command()
+    @click.argument(
+        "method", type=click.Choice(["not-none", None], case_sensitive=False)
+    )
+    def cli(method: str | None):
+        assert isinstance(method, str) or method is None
+        click.echo(method)
+
+    result = runner.invoke(cli, ["not-none"])
+    assert not result.exception
+    assert result.output == "not-none\n"
+
+    # None is not yet supported.
+    result = runner.invoke(cli, ["none"])
+    assert result.exception
 
 
 def test_datetime_option_default(runner):

--- a/tests/test_formatting.py
+++ b/tests/test_formatting.py
@@ -248,7 +248,7 @@ def test_formatting_usage_custom_help(runner):
 
 def test_formatting_custom_type_metavar(runner):
     class MyType(click.ParamType):
-        def get_metavar(self, param):
+        def get_metavar(self, param: click.Parameter, ctx: click.Context):
             return "MY_TYPE"
 
     @click.command("foo")

--- a/tests/test_info_dict.py
+++ b/tests/test_info_dict.py
@@ -106,11 +106,11 @@ HELLO_GROUP = (
         ),
         pytest.param(*STRING_PARAM_TYPE, id="STRING ParamType"),
         pytest.param(
-            click.Choice(["a", "b"]),
+            click.Choice(("a", "b")),
             {
                 "param_type": "Choice",
                 "name": "choice",
-                "choices": ["a", "b"],
+                "choices": ("a", "b"),
                 "case_sensitive": True,
             },
             id="Choice ParamType",

--- a/tests/test_normalization.py
+++ b/tests/test_normalization.py
@@ -17,12 +17,37 @@ def test_option_normalization(runner):
 
 def test_choice_normalization(runner):
     @click.command(context_settings=CONTEXT_SETTINGS)
-    @click.option("--choice", type=click.Choice(["Foo", "Bar"]))
-    def cli(choice):
-        click.echo(choice)
+    @click.option(
+        "--method",
+        type=click.Choice(
+            ["SCREAMING_SNAKE_CASE", "snake_case", "PascalCase", "kebab-case"],
+            case_sensitive=False,
+        ),
+    )
+    def cli(method):
+        click.echo(method)
 
-    result = runner.invoke(cli, ["--CHOICE", "FOO"])
-    assert result.output == "Foo\n"
+    result = runner.invoke(cli, ["--METHOD=snake_case"])
+    assert not result.exception, result.output
+    assert result.output == "snake_case\n"
+
+    # Even though it's case sensitive, the choice's original value is preserved
+    result = runner.invoke(cli, ["--method=pascalcase"])
+    assert not result.exception, result.output
+    assert result.output == "PascalCase\n"
+
+    result = runner.invoke(cli, ["--method=meh"])
+    assert result.exit_code == 2
+    assert (
+        "Invalid value for '--method': 'meh' is not one of "
+        "'screaming_snake_case', 'snake_case', 'pascalcase', 'kebab-case'."
+    ) in result.output
+
+    result = runner.invoke(cli, ["--help"])
+    assert (
+        "--method [screaming_snake_case|snake_case|pascalcase|kebab-case]"
+        in result.output
+    )
 
 
 def test_command_normalization(runner):

--- a/tests/test_types.py
+++ b/tests/test_types.py
@@ -248,5 +248,5 @@ def test_invalid_path_with_esc_sequence():
 
 def test_choice_get_invalid_choice_message():
     choice = click.Choice(["a", "b", "c"])
-    message = choice.get_invalid_choice_message("d")
+    message = choice.get_invalid_choice_message("d", ctx=None)
     assert message == "'d' is not one of 'a', 'b', 'c'."

